### PR TITLE
[docs-only] add toc to changelog

### PR DESCRIPTION
**Pull request by mmattel**
## Description

Adds table of contents to the top of the CHANGELOG.md file.

### Anchor links in markdown are generated from the content of the header according to the following rules:

1. All text is converted to lowercase.
2. All non-word text (e.g., punctuation, HTML) is removed.
3. All spaces are converted to hyphens.
4. Two or more hyphens in a row are converted to one.
5. If a header with the same ID has already been generated, a unique incrementing number is appended, starting at 1.
6. All changelogs for 2.x and below have a main entrance link

## Related Issue
- Relates to https://github.com/owncloud/docs/issues/4901
- Changes copied from: https://github.com/owncloud/ocis/pull/7691

### CODE REVIEW

File: changelog/CHANGELOG.tmpl [hunk 1]
```diff
@@ -1,7 +1,12 @@
+# Table of Contents
+
+{{ range . -}}
+  * [Changelog for {{ .Version }}](#changelog-for-{{ .Version | replace "." ""}}-{{ .Date -}})
+{{ end -}}
 {{ $allVersions := . }}
```
Comment (mmattel) [inline changelog/CHANGELOG.tmpl:6]:
```suggestion
{{ end -}}
  * [ChangeLog for 2.x versions](#ChangeLog-for-2x-versions)
{{ $allVersions := . }}
```

**Comment by michaelstingl** [comment]
@mmattel could you post the desktop TOC  to get an idea?

File: changelog/CHANGELOG.tmpl [hunk 2]
```diff
@@ -1,7 +1,13 @@
+# Table of Contents
+
+{{ range . -}}
+  * [Changelog for {{ .Version }}](#changelog-for-{{ .Version | replace "." ""}}-{{ .Date -}})
+{{ end -}}
+  * [ChangeLog for 2.x versions](#ChangeLog-for-2x-versions)
```
Comment (mmattel) [inline changelog/CHANGELOG.tmpl:6]:
```suggestion
  * [ChangeLog for 2.x versions and below](#ChangeLog-for-2x-versions-and-below)
```

File: changelog/CHANGELOG.tmpl [hunk 3]
```diff
@@ -47,8 +51,7 @@ Details
 {{ end }}{{ end -}}
 
 {{/* Start of old changelog */ -}}
-ChangeLog
-=========
+# ChangeLog for 2.x versions
```
Comment (mmattel) [inline changelog/CHANGELOG.tmpl:54]:
```suggestion
# ChangeLog for 2.x versions and below
```

**Comment by mmattel** [comment]
Run locally, works nicely 😄 
image

File: changelog/CHANGELOG.tmpl [hunk 4]
```diff
@@ -1,7 +1,13 @@
...
+
+{{ range . -}}
+  * [Changelog for {{ .Version }}](#changelog-for-owncloud-desktop-client-{{ .Version | replace "." ""}}-{{ .Date | lower -}})
+{{ end -}}
+  * [Changelog for 2.6.1 versions and below](#changelog-for-261-versions-and-below)
 {{ $allVersions := . }}
 {{- range $index, $changes := . }}{{ with $changes -}}
-Changelog for ownCloud Desktop Client [{{ .Version }}] ({{ .Date }})
-=======================================
+# Changelog for ownCloud Desktop Client [{{ .Version }}] ({{ .Date }})
```
Comment (fmoc) [inline changelog/CHANGELOG.tmpl:9]: You should add one or two blank lines above this heading. Right now, rendering may or may not work. Also, the resulting generated code is kind of hard to read.

```suggestion


# Changelog for ownCloud Desktop Client [{{ .Version }}] ({{ .Date }})
```

File: changelog/CHANGELOG.tmpl [hunk 5]
```diff
@@ -1,7 +1,13 @@
+# Table of Contents
```
Comment (fmoc) [inline changelog/CHANGELOG.tmpl:1]: I'd suggest a third level heading.

**Review by fmoc — APPROVED** [review]

**Review by fmoc — CHANGES_REQUESTED** [review]
See above.

File: changelog/CHANGELOG.tmpl (hunk 4 above)
Comment (mmattel) [reply changelog/CHANGELOG.tmpl:9]: I can tell that we have that in all repos where we use this TOC generating code the same and it works without any issues. The code was created by mbarz and I have tested it successfully locally using calens. I would therefore keep that as it is to have that harmonized with the other templates.

File: changelog/CHANGELOG.tmpl (hunk 5 above)
Comment (fmoc) [reply changelog/CHANGELOG.tmpl:1]: Fine by me...

File: changelog/CHANGELOG.tmpl (hunk 4 above)
Comment (fmoc) [reply changelog/CHANGELOG.tmpl:9]: Should be taken into account upon the next update then, I guess. The generated code is still pretty ugly.

**Review by fmoc — APPROVED** [review]

**Pull request merged**